Add support for AmgX

artv3 opened the pull request:
Building AmgX  +  MFEM 
``bash ./setupAMGXMFEM.sh``

Sample runs: 

AmgX as preconditioner: 
``lrun -n 1  --smpiargs="off" ./amgx_ex1 --amgx-file precon.json --amgx-preconditioner  ``                                            

AmgX as solver
``lrun -n 1  --smpiargs="off"    ./amgx_ex1 --amgx-file multi_gs.json --amgx-solver   ``

MPI Teams sharing a GPU:
``mpirun -n 40 ./amgx_ex1p --amgx-file amg_pcg.json ``

Map 1 MPI rank to a GPU: 
``lrun -n 4 ./amgx_ex1p --amgx-file amg_pcg.json --amgx-mpi-gpu-exclusive ``



<!--GHEX{"id":1386,"author":"artv3","editor":"tzanio","reviewers":["dylan-copeland","acfisher","v-dobrev","barker29"],"assignment":"2020-09-08T12:47:20-07:00","approval":"2020-10-08T19:09:58.476Z","merge":"2020-10-18T18:19:40.488Z"}XEHG--><!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1386](https://github.com/mfem/mfem/pull/1386) | @artv3 | @tzanio | @dylan-copeland + @acfisher + @v-dobrev + @barker29 | 09/08/20 | 10/08/20 | 10/18/20 | |
<!--ELBATXEHG-->

barker29 commented:
Have you built AMGX on lassen? I'm getting errors, in my case it looks like it's trying to pick up gcc includes even though I'm trying to build with xlc.

This is probably just my own mistake, but if you have a recipe for building on lassen I'd like to see it.
```
In file included from /g/g14/barker29/ws/packages/AMGX/core/src/aggregation/aggregation_amg_level.cu:28:
In file included from /g/g14/barker29/ws/packages/AMGX/core/include/aggregation/aggregation_amg_level.h:40:
In file included from /g/g14/barker29/ws/packages/AMGX/core/../base/include/amg_level.h:34:
In file included from /g/g14/barker29/ws/packages/AMGX/core/../base/include/amg.h:53:
In file included from /g/g14/barker29/ws/packages/AMGX/core/../base/include/amg_solver.h:45:
In file included from /g/g14/barker29/ws/packages/AMGX/core/../base/include/vector.h:81:
In file included from /g/g14/barker29/ws/packages/AMGX/core/../base/include/resources.h:36:
In file included from /g/g14/barker29/ws/packages/AMGX/core/../base/include/thread_manager.h:35:
In file included from /g/g14/barker29/ws/packages/AMGX/core/../base/include/global_thread_handle.h:34:
In file included from /usr/tce/packages/gcc/gcc-4.9.3/gnu/lib64/gcc/powerpc64le-unknown-linux-gnu/4.9.3/../../../../include/c++/4.9.3/mutex:35:
/usr/tce/packages/gcc/gcc-4.9.3/gnu/lib64/gcc/powerpc64le-unknown-linux-gnu/4.9.3/../../../../include/c++/4.9.3/bits/c++0x_warning.h:32:2: error: This file requires compiler and library support for the ISO C++ 2011 standard. This support is currently experimental, and must be enabled with the -std=c++11 or -std=gnu++11 compiler options.
#error This file requires compiler and library support for the \
 ^
CMake Error at amgx_core_generated_aggregation_amg_level.cu.o.Release.cmake:226 (message):
  Error generating
  /g/g14/barker29/ws/packages/AMGX/build/core/CMakeFiles/amgx_core.dir/src/aggregation/./amgx_core_generated_aggregation_amg_level.cu.o


make[2]: *** [core/CMakeFiles/amgx_core.dir/build.make:65: core/CMakeFiles/amgx_core.dir/src/aggregation/amgx_core_generated_aggregation_amg_level.cu.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:259: core/CMakeFiles/amgx_core.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
```

artv3 commented:
> Have you built AMGX on lassen? I'm getting errors, in my case it looks like it's trying to pick up gcc includes even though I'm trying to build with xlc.
> 
> This is probably just my own mistake, but if you have a recipe for building on lassen I'd like to see it.
> 
> ```
> In file included from /g/g14/barker29/ws/packages/AMGX/core/src/aggregation/aggregation_amg_level.cu:28:
> In file included from /g/g14/barker29/ws/packages/AMGX/core/include/aggregation/aggregation_amg_level.h:40:
> In file included from /g/g14/barker29/ws/packages/AMGX/core/../base/include/amg_level.h:34:
> In file included from /g/g14/barker29/ws/packages/AMGX/core/../base/include/amg.h:53:
> In file included from /g/g14/barker29/ws/packages/AMGX/core/../base/include/amg_solver.h:45:
> In file included from /g/g14/barker29/ws/packages/AMGX/core/../base/include/vector.h:81:
> In file included from /g/g14/barker29/ws/packages/AMGX/core/../base/include/resources.h:36:
> In file included from /g/g14/barker29/ws/packages/AMGX/core/../base/include/thread_manager.h:35:
> In file included from /g/g14/barker29/ws/packages/AMGX/core/../base/include/global_thread_handle.h:34:
> In file included from /usr/tce/packages/gcc/gcc-4.9.3/gnu/lib64/gcc/powerpc64le-unknown-linux-gnu/4.9.3/../../../../include/c++/4.9.3/mutex:35:
> /usr/tce/packages/gcc/gcc-4.9.3/gnu/lib64/gcc/powerpc64le-unknown-linux-gnu/4.9.3/../../../../include/c++/4.9.3/bits/c++0x_warning.h:32:2: error: This file requires compiler and library support for the ISO C++ 2011 standard. This support is currently experimental, and must be enabled with the -std=c++11 or -std=gnu++11 compiler options.
> #error This file requires compiler and library support for the \
>  ^
> CMake Error at amgx_core_generated_aggregation_amg_level.cu.o.Release.cmake:226 (message):
>   Error generating
>   /g/g14/barker29/ws/packages/AMGX/build/core/CMakeFiles/amgx_core.dir/src/aggregation/./amgx_core_generated_aggregation_amg_level.cu.o
> 
> 
> make[2]: *** [core/CMakeFiles/amgx_core.dir/build.make:65: core/CMakeFiles/amgx_core.dir/src/aggregation/amgx_core_generated_aggregation_amg_level.cu.o] Error 1
> make[1]: *** [CMakeFiles/Makefile2:259: core/CMakeFiles/amgx_core.dir/all] Error 2
> make: *** [Makefile:130: all] Error 2
> ```

I was able to build on RzAnsel.  I didn't try specifying a compiler though, I just let CMAKE pull defaults.  Perhaps try doing module load gcc/7.3.1 or try adding the following flag -std=c++11? 

How did you configure the build?

barker29 commented:
I'm specifically trying to avoid using gcc because the docs for lassen discourage it and the rest of my stack (metis, hypre, mfem) is built with xlc. But so far at least the AMGX build is behaving much better with gcc so we'll see how it goes.

artv3 commented:
@barker29 , I was able to build on lassen with the following cmake config: 
`` cmake -DCMAKE_CXX_COMPILER=xlC -DCMAKE_C_COMPILER=xlc -DCMAKE_CXX_FLAGS=-std=c++11 ../`` . I did get lots of warnings though. I can relay them to the AMGX team.

barker29 commented:
Thanks! I was missing the `std=c++11` bit, which is obvious now.

barker29 commented:
Can you post the expected output from `./amgx_ex -c ../../AMGX/core/configs/CLASSICAL_V_CYCLE.json`?

I am getting something that seems to maybe work, but then I get an error from some `CuMemFree` call and I want to see if what is happening before that is reasonable or not.

artv3 commented:
> Can you post the expected output from `./amgx_ex -c ../../AMGX/core/configs/CLASSICAL_V_CYCLE.json`?
> 
> I am getting something that seems to maybe work, but then I get an error from some `CuMemFree` call and I want to see if what is happening before that is reasonable or not.

```
The output I get is: 
AMGX version 2.1.0.131-opensource
Built on Apr  7 2020, 11:55:12
Compiled with CUDA Runtime 10.1, using CUDA driver 10.1
Options used:
   --mesh ../data/beam-hex.mesh
   --order 1
   --no-static-condensation
   --no-partial-assembly
   --c ../../AMGX/core/configs/CLASSICAL_V_CYCLE.json
   --device cuda
   --visualization
Device configuration: cuda,cpu
Memory configuration: host-std,cuda
Number of finite element unknowns: 37281
AMG Grid:
         Number of Levels: 1
            LVL         ROWS               NNZ    SPRSTY       Mem (GB)
         --------------------------------------------------------------
           0(D)        37281            924385  0.000665         0.0109
         --------------------------------------------------------------
         Grid Complexity: 1
         Operator Complexity: 1
         Total Memory Usage: 0.0108864 GB
         --------------------------------------------------------------
           iter      Mem Usage (GB)       residual           rate
         --------------------------------------------------------------
            Ini             12.2691   4.272940e-02
Relative residual has reached machine precision
              0             12.2691   6.177668e-18         0.0000
         --------------------------------------------------------------
         Total Iterations: 1
         Avg Convergence Rate: 		         0.0000
         Final Residual: 		   6.177668e-18
         Total Reduction in Residual: 	   1.445765e-16
         Maximum Memory Usage: 		         12.269 GB
         --------------------------------------------------------------
Total Time: 6.92136
    setup: 6.89951 s
    solve: 0.0218459 s
    solve(per iteration): 0.0218459 s
```

Is this the error you got?
```
CUDA Hook Library: Failed to find symbol mem_find_dreg_entries, ./amgx_ex: undefined symbol: __PAMI_Invalidate_region
CUDA error: (cudaFree(dptr)) failed with error:
 --> driver shutting down
 ... in function: void *mfem::CuMemFree(void *)
 ... in file: general/cuda.cpp:71
ERROR:  One or more process (first noticed rank 0) terminated with signal 6
```
I suspect you ran the code the following way: 
``lrun -n 1 ./amgx_ex -c ../../AMGX/core/configs/CLASSICAL_V_CYCLE.json``

Try the following: 
``lrun -n 1 --smpiargs="off" ./amgx_ex -c ../../AMGX/core/configs/CLASSICAL_V_CYCLE.json``

barker29 commented:
Beautiful, thank you!

barker29 commented:
I just pushed a commit that begins to imagine an MFEM object that interfaces to AMGX. It is still very much WIP but I think it's an improvement. @artv3 if you don't like it, feel free to roll it back (or ask me to). I'd appreciate feedback from anyone.

artv3 commented:
@barker29 Nice! I think that looks fantastic! I also found this AMGX PETSc wrapper: https://github.com/barbagroup/AmgXWrapper. By creating PETSc matrices through MFEM, I've been able to do multi-GPU runs. I can push up an example of how I've made it work this week. 

Within MFEM, I think we can do something similar and bypass the need of PETSc.

artv3 commented:
@barker29 , I added a first pass for multi-GPU support, see ex1p. It works by assuming 1 GPU per MPI proc. I plan to do a clean up pass and gather some performance results. It's looking promising so far, it would be great if we can merge our wrappers together as well.

artv3 commented on examples/ex1p.cpp:
```diff
@@ -65,11 +65,15 @@ int main(int argc, char *argv[])
    bool static_cond = false;
    bool pa = false;
+   bool amgx = true;
    const char *device_config = "cpu";
```
During testing, I left the MFEM device set to CPU. I think AmgX has a way to identify if the memory is the host or device and do the transfers accordingly. Something, I'd like to inquire about.

barker29 replied:
I'm interested in this too! Based on just a few runs it seems to produce identical output with device `cuda` and `cpu`, I'm curious what it's actually doing.

artv3 commented:
I've been doing some run-time comparisons using AMGX and  Boomer AMG + PCG. The driver is perf_amgx.cpp and amg_pcg.json is the config file for AMGX (I think that should be a similar configuration to boomer amg + pcg). 

Below are some performance numbers using 1 node of Lassen.  The total time includes any sparse matrix  manipulation required for the solvers. For AMGX I have to consolidate the diagonal and off diagonal sparse matrices from the Hypre parcsr format, I'm wondering if this step could could be done on the GPU to further improve performance?

AMGX - Order 1 
| GPU  | Dofs | Setup + Solve (sec) | Total time (sec) |
| ------------- | ------------- |------------- |------------- |
| 1  | 2146689 | 2.82216 | 5.48312 |
| 2 | 4276737  |  6.31895 | 9.21916   |
| 4 | 8520321  | 3.26739  |  6.51875 |

Boomer AMG + PCG - Order 1
| MPI Procs  | Dofs | Setup + Solve (sec) | Total time (sec) |
| ------------- | ------------- |------------- |------------- |
| 1  | 545025  | 6.89 | 6.89136 |
| 2 | 1081665  | 7.37  |  7.36441  |
| 4 | 2146689  |  6.98 |  6.97845 |
| 8 | 4276737  |  8.52 | 8.51148  |
| 16 |  8520321 |  8.8 | 8.80561  |
| 32   | 16974593  | 9.08  | 9.08053 |

barker29 commented:
Those are nice results. I'd be curious to see what happens at higher order - it's possible the GPU looks better because there's more operations to do, but the access patterns aren't regular so it might not matter.

artv3 commented:
> Those are nice results. I'd be curious to see what happens at higher order - it's possible the GPU looks better because there's more operations to do, but the access patterns aren't regular so it might not matter.

I'm curious as well, I'll post some more results later this evening. I'll try higher order + more GPUs.

artv3 commented:
Additional runs below. To improve performance we could try CUDA pinned memory and see if we can skip the sparse matrix wrangling (or do it on the device). 


Order 3 | AMGX |   |   | Boomer+PCG |   |  
-- | -- | -- | -- | -- | -- | --
GPUs | Dofs | Setup+Solve | Total Time | MPI | Dofs | Setup+Solve | Total Time
1 | 912673 | 2.08512 | 7.43861 | 1 | 117649 | 6.04 | 6.03737
2 | 1815937 | 3.76649 | 9.29254 | 2 | 232897 | 6.94 | 6.94391
4 | 3613153 | 3.60117 | 9.21555 | 4 | 461041 | 8.72 | 8.71538
8 | 7189057 | 3.16233 | 8.98515 | 8 | 912673 | 7.04 | 7.03111
16 | 14340865 | 5.63106 | 11.7472 | 16 | 1815937 | 8.21 | 8.20503
32 | 28607425 | 6.43755 | 13.1838 | 32 | 3613153 | 9.96 | 9.9573
  |   |   |   | 64 | 7189057 | 7.73 | 7.71505
  |   |   |   | 128 | 14340865 | 9.46 | 9.45479
  |   |   |   | 256 | 28607425 | 11.64 | 11.636


Order 4 | AMGX |   |   | Boomer+PCG |   |  
-- | -- | -- | -- | -- | -- | --
GPUs | Dofs | Setup+Solve | Total Time | MPI | Dofs | Setup+Solve | Total Time
1 | 274625 | 2.78495 | 5.80421 | 1 | 35937 | 2.62 | 2.61288
2 | 545025 | 3.15463 | 6.32125 | 2 | 70785 | 3.6 | 3.59559
4 | 1081665 | 3.18324 | 6.47475 | 4 | 139425 | 4.01 | 4.00496
8 | 2146689 | 6.0769 | 9.55185 | 8 | 274625 | 3.72 | 3.70214
16 | 4276737 | 4.33204 | 8.05218 | 16 | 545025 | 4.44 | 4.43781
32 | 8520321 | 4.81449 | 8.87212 | 32 | 1081665 | 5.04 | 5.02581
  |   |   |   | 64 | 2146689 | 4.59 | 4.57185
  |   |   |   | 128 | 4276737 | 5.35 | 5.34299
  |   |   |   | 256 | 8520321 | 5.96 | 5.95573

v-dobrev commented:
Somehow there is a relatively big difference between orders 3 and 4:
```ascii
order 3, 32 GPUs vs 256 CPUs (28607425 dofs): 6.43755 vs 11.64 setup+solve time
order 4, 32 GPUs vs 256 CPUs ( 8520321 dofs): 4.81449 vs  5.96 setup+solve time
```
This is probably due to the smaller local problem size for order 4.

Overall, AMGX gives nice boost but not as big as one might hope for.

It'll be interesting to compare 32 GPUs vs 256 CPUs at order 1. The results above for order 1 use 1 GPU to 4 CPUs ratio which is not quite "fair".

By the way, is this 2D or 3D?

artv3 commented:
@v-dobrev , these are 3D results using the perf_amgx.cpp driver. 

I just updated new results since I made the mistake of not monitoring the residual, and some problems didn't converge. I pushed in a new configuration where I observed convergence across the board. Additionally, I am now pinning the host buffers using ``AMGX_pin_memory``. 

There are a few differences I should also highlight.
Although both solvers use AMG as a preconditioner, and PCG to solve the system. AMGX uses Jacobi as a smoother. Boomer is configured with Gauss-Siedel, I tried using colored Gauss–Seidel  in  AMGX but it was much more expensive (sequential Gauss-Siedel is not suited for the GPU). 
PMIS is also used as selector in AMGX vs HMIS in Boomer.

artv3 commented:
@am-oler when you get a chance can you merge with master and make sure we pass on continuous integration and travis?

artv3 commented on config/defaults.mk:
```diff
@@ -374,7 +374,7 @@ UMPIRE_OPT = -I$(UMPIRE_DIR)/include
 
 # AMGX library configuration
-AMGX_DIR = @MFEM_DIR@/../amgx
+AMGX_DIR = @MFEM_DIR@/../AMGX
```
@am-oler could you leave as amgx (lower case)? The CEED benchmarks build packages using lowercase names.

artv3 commented on linalg/AmgXSolver.hpp:
```diff
@@ -0,0 +1,230 @@
+
+  virtual void SetOperator(const Operator &op);
+
+  virtual void SetOperator(const SparseMatrix &in_A);
```
Here you should be able to only have 1 set operator method that takes in an Operator type. Then depending on what is passed in do the right thing (there should be a way to check what type of object is passed in).

artv3 commented on examples/driver.cpp:
```diff
@@ -0,0 +1,316 @@
+//                       MFEM Example 1 - Parallel Version
+//
+// Compile with: make ex1p
```
If this is not needed, please delete.

artv3 commented on examples/perf_amgx.cpp:
```diff
@@ -0,0 +1,322 @@
+   args.AddOption(&amgx_cfg, "-c","--c","AMGX solver file");
+   args.AddOption(&amgx, "-amgx","--amgx","-no-amgx",
+                  "--no-amgx","Use AMGX");
+   args.Parse();
```
Please delete this since we now have amgx working with the CEED benchmarks.

artv3 commented on examples/amgx_ex4p.cpp:
```diff
@@ -0,0 +1,385 @@
+      amgx_str = amgx_cfg;
+      AmgXSolver amgx;
+
+      amgx.initialize(MPI_COMM_WORLD, "dDDI", amgx_str, ndevices);
```
@am-oler , I think you can support lrun if you have ndevices=1 by default. So like that 1 MPI proc only sees 1 GPU.

artv3 commented:
Hooray! Passed on travis!

artv3 commented:
@tzanio @v-dobrev , I have a few smaller TODO's left but I feel the code is ready for some eyes if folks are interested in trying it out. Any feedback on further organization and features would be appreciated. I updated the PR description with some sample runs.

dylan-copeland commented:
@artv3 Do you plan to keep the example files in the examples directory in master, or are they just in the PR for demonstration to the reviewers?

artv3 commented:
> @artv3 Do you plan to keep the example files in the examples directory in master, or are they just in the PR for demonstration to the reviewers?

I originally added them for the reviewers, but perhaps we can have a folder with an AmgX example? I will follow the preference of the reviewers.

dylan-copeland commented:
@artv3 The first sample run `lrun -n 1 --smpiargs="off" amgx_ex --amgx-file MULTI_GS.json --amgx-solver` works for me, but the second one does not:
`lrun -n 1 --smpiargs="off" amgx_ex --amgx-file precon.json --amgx-preconditioner
Options used:
   --mesh ../data/star.mesh
   --order 1
   --no-static-condensation
   --no-partial-assembly
   --amgx-file precon.json
   --amgx-preconditioner
   --amgx-verbose
   --device cpu
   --visualization
Device configuration: cpu
Memory configuration: host-std
Number of finite element unknowns: 20801
AMGX version 2.1.0.131-opensource
Built on Sep 16 2020, 12:41:55
Compiled with CUDA Runtime 10.1, using CUDA driver 10.1
Applying AmgX as preconditioner 
   Iteration :   0  (B r, r) = 1.1207e+231 ...
PCG: The operator is not positive definite. (Ad, d) = nan
   Iteration :  40  (B r, r) = -nan
PCG: No convergence!
Average reduction factor = -nan`

artv3 commented:
> @artv3 The first sample run `lrun -n 1 --smpiargs="off" amgx_ex --amgx-file MULTI_GS.json --amgx-solver` works for me, but the second one does not:
> 
> ```
> lrun -n 1 --smpiargs="off" amgx_ex --amgx-file precon.json --amgx-preconditioner
> Options used:
>    --mesh ../data/star.mesh
>    --order 1
>    --no-static-condensation
>    --no-partial-assembly
>    --amgx-file precon.json
>    --amgx-preconditioner
>    --amgx-verbose
>    --device cpu
>    --visualization
> Device configuration: cpu
> Memory configuration: host-std
> Number of finite element unknowns: 20801
> AMGX version 2.1.0.131-opensource
> Built on Sep 16 2020, 12:41:55
> Compiled with CUDA Runtime 10.1, using CUDA driver 10.1
> Applying AmgX as preconditioner 
>    Iteration :   0  (B r, r) = 1.1207e+231 ...
> PCG: The operator is not positive definite. (Ad, d) = nan
>    Iteration :  40  (B r, r) = -nan
> PCG: No convergence!
> Average reduction factor = -nan
> ```
> 
> Another issue is that the first sample run works for `MassIntegrator`, but if this is changed to `DiffusionIntegrator` as in `ex1p` then it fails to converge.

Sorry about that, it seems we have to set X=0 before we apply mult in AMGX (just pushed a fix). 
I now get the following: 

```
[vargas45@rzansel30:examples]$  lrun -n 1 --smpiargs="off" ./amgx_ex --amgx-file precon.json --amgx-preconditioner
Compiled with CUDA Runtime 10.1, using CUDA driver 10.2
Applying AmgX as preconditioner 
   Iteration :   0  (B r, r) = 4.78951 ...
   Iteration :   6  (B r, r) = 3.78895e-13
```

With regards to solving with the diffusion matrix, I see that too. I see improved performance with amg as a preconditioner + cg (see below) I suspect we have to adjust the multi-color GS script for better performance. 
```
 lrun -n 1 --smpiargs="off" ./amgx_ex --amgx-file amg_pcg.json --amgx-solver
```

AMG Grid:
         Number of Levels: 3
            LVL         ROWS               NNZ    SPRSTY       Mem (GB)
         --------------------------------------------------------------
           0(D)        20801            185281  0.000428        0.00259
           1(D)         3924             84758    0.0055        0.00205
           2(D)          707             20697    0.0414       0.000484
         --------------------------------------------------------------
         Grid Complexity: 1.22263
         Operator Complexity: 1.56916
         Total Memory Usage: 0.00512902 GB
         --------------------------------------------------------------
Applying AmgX as solver 
           iter      Mem Usage (GB)       residual           rate
         --------------------------------------------------------------
            Ini            0.878906   3.309898e-02
              0            0.878906   8.023642e-01        24.2414
              1              0.8789   1.107793e-01         0.1381
              2              0.8789   1.980708e-02         0.1788
              3              0.8789   3.377608e-03         0.1705
              4              0.8789   5.967222e-04         0.1767
              5              0.8789   1.052048e-04         0.1763
              6              0.8789   1.834787e-05         0.1744
              7              0.8789   3.205434e-06         0.1747
              8              0.8789   5.627631e-07         0.1756
              9              0.8789   9.496952e-08         0.1688
             10              0.8789   1.610142e-08         0.1695
             11              0.8789   2.777073e-09         0.1725
             12              0.8789   4.709810e-10         0.1696
             13              0.8789   8.258003e-11         0.1753
             14              0.8789   1.435589e-11         0.1738
             15              0.8789   2.500842e-12         0.1742
             16              0.8789   4.359400e-13         0.1743
         --------------------------------------------------------------
         Total Iterations: 17
         Avg Convergence Rate: 		         0.2291
         Final Residual: 		   4.359400e-13
         Total Reduction in Residual: 	   1.317080e-11
         Maximum Memory Usage: 		          0.879 GB
         ------------------------------------------------------------

dylan-copeland commented on linalg/AmgXSolver.hpp:
```diff
@@ -0,0 +1,228 @@
+// Copyright (c) 2010-2020, Lawrence Livermore National Security, LLC. Produced
```
This hpp and the corresponding cpp file would be the first to begin with a capital letter. Maybe they should be renamed in lower-case.

dylan-copeland commented on linalg/AmgXSolver.cpp:
```diff
@@ -0,0 +1,846 @@
+      MPI_Comm_size(gpuWorld, &gpuWorldSize);
+      MPI_Comm_rank(gpuWorld, &myGpuWorldRank);
+   }
+   else // for those can not communicate with GPU devices
```
those that?

dylan-copeland commented on linalg/AmgXSolver.cpp:
```diff
@@ -0,0 +1,846 @@
+{
+
+   // Set the ID of device that each local process will use
+   if (nDevs == localSize) // # of the devices and local precosses are the same
```
processes

dylan-copeland commented on linalg/AmgXSolver.cpp:
```diff
@@ -0,0 +1,846 @@
+               outArr.HostWrite(), Apart.HostRead(), Adisp.HostRead(),
+               MPI_DOUBLE, 0, mpiTeamComm);
+}
+
```
There are many extra empty lines in this file.

dylan-copeland commented on linalg/AmgXSolver.cpp:
```diff
@@ -0,0 +1,846 @@
+      SetA(*Aptr);
+   }
+#endif
+}
```
`else mfem_error`?

dylan-copeland commented:
> > @artv3 Do you plan to keep the example files in the examples directory in master, or are they just in the PR for demonstration to the reviewers?
> 
> I originally added them for the reviewers, but perhaps we can have a folder with an AmgX example? I will follow the preference of the reviewers.

I definitely think we need at least one example to demonstrate how to use `AmgXSolver`. It seems it could be added to the main `examples/ex1p` with `#ifdef MFEM_USE_AMGX`, without adding very many lines or confusing the reader. I know the `examples` directory should be kept as simple as possible, and maybe other reviewers or editors have stronger opinions, but it seems to me that just a few lines of code could enable the option to use AmgX. Otherwise, I would also be fine with an example or two in another directory.

dylan-copeland commented on linalg/AmgXSolver.cpp:
```diff
@@ -0,0 +1,846 @@
+
+void AmgXSolver::Initialize_Serial(const std::string &modeStr,
+                                   const std::string &cfgFile)
+{
```
Is it necessary to specify the configuration parameters by an input file? This could be inconvenient and restrictive.	It would give the user more flexibility to have a constructor allowing specification of the parameters in the code. Would that be possible with AmgX?

acfisher commented on examples/amgx_ex.cpp:
```diff
@@ -0,0 +1,269 @@
+      }
+      // Recover the solution as a finite element grid function.
+      a->RecoverFEMSolution(X, *b, x);
+      //x.HostReadWrite();
```
Dead code?

acfisher commented on examples/amgx_exclusive_ex1p.cpp:
```diff
@@ -0,0 +1,316 @@
+
+   // 2. Parse command-line options.
+   const char *mesh_file = "../data/star.mesh";
+   //const char *mesh_file = "../data/beam-hex.mesh";
```
Dead code?

acfisher commented on examples/amgx_exclusive_ex1p.cpp:
```diff
@@ -0,0 +1,316 @@
+   //    this example we do 'ref_levels' of uniform refinement. We choose
+   //    'ref_levels' to be the largest number that gives a final mesh with no
+   //    more than 10,000 elements.
+   {
```
Any reason for the extra block here?

barker29 commented:
Can we update the TODOs at the very top of this PR? I suspect some of them are complete but I would like that confirmed.

artv3 commented:
Thanks all, I'll push on this PR this week.

barker29 commented:
In addition to comments from others above, I'm trying to run some examples on this branch on lassen and generally having some trouble. For example:

- I can't build `amgx_ex`:
```
amgx_ex.cpp(193): error: identifier "AmgXSolver" is undefined
```
- `./amgx_ex4p -d cuda` does not seem to run on device:
```
Verification failed: (Device::IsDisabled()) is false:
 --> transpose action on device is not enabled; see BuildTranspose() for details.
 ... in function: virtual void mfem::SparseMatrix::AddMultTranspose(const mfem::Vector &, mfem::Vector &, const double) const
 ... in file: linalg/sparsemat.cpp:759
```
- `./amgx_ex_orig -d cuda` seems to run and converge, but has some warnings/errors:
```
CUDA Hook Library: Failed to find symbol mem_find_dreg_entries, ./amgx_ex_orig: undefined symbol: __PAMI_Invalidate_region
```

In general it would be nice to have cleaner examples with example command lines so I know what exactly to run and what output to expect.

artv3 commented:
> In addition to comments from others above, I'm trying to run some examples on this branch on lassen and generally having some trouble. For example:
> 
>     * I can't build `amgx_ex`:
> 
> 
> ```
> amgx_ex.cpp(193): error: identifier "AmgXSolver" is undefined
> ```
> 
>     * `./amgx_ex4p -d cuda` does not seem to run on device:
> 
> 
> ```
> Verification failed: (Device::IsDisabled()) is false:
>  --> transpose action on device is not enabled; see BuildTranspose() for details.
>  ... in function: virtual void mfem::SparseMatrix::AddMultTranspose(const mfem::Vector &, mfem::Vector &, const double) const
>  ... in file: linalg/sparsemat.cpp:759
> ```
> 
>     * `./amgx_ex_orig -d cuda` seems to run and converge, but has some warnings/errors:
> 
> 
> ```
> CUDA Hook Library: Failed to find symbol mem_find_dreg_entries, ./amgx_ex_orig: undefined symbol: __PAMI_Invalidate_region
> ```
> 
> In general it would be nice to have cleaner examples with example command lines so I know what exactly to run and what output to expect.

Hi @barker29, 

I just built the whole stack using the bash script in the description of the PR. I was able to build amgx_ex. Are you up to date with the PR? I can look into it with you if it doesn't work. 

Did you also try some of the sample runs on the PR description: 
For example: 

 mpirun -n 40 ./amgx_ex4p --amgx-file amg_pcg.json -nd 1 
Gives the following output: 

```[vargas45@rzansel28:examples]$  mpirun -n 40 ./amgx_ex4p --amgx-file amg_pcg.json -nd 1 
Options used:
   --mesh ../data/beam-hex.mesh
   --order 1
   --impose-bc
   --frequency 1
   --no-static-condensation
   --hybridization
   --no-partial-assembly
   --device cpu
   --visualization
   --amgx-file amg_pcg.json
   --amgx
   --nr 2
   --np 2
   --nd 1
   --nsolves 1
Device configuration: cpu
Memory configuration: host-std
Number of finite element unknowns: 102656
AMGX version 2.1.0.131-opensource
Built on Sep 24 2020, 09:42:26
Compiled with CUDA Runtime 10.1, using CUDA driver 10.2
Cannot read file as JSON object, trying as AMGX config
Converting config string to current config version
Parsing configuration string: exception_handling=1 ; 
Using Normal MPI (Hostbuffer) communicator...
AMG Grid:
         Number of Levels: 4
            LVL         ROWS               NNZ    SPRSTY       Mem (GB)
         --------------------------------------------------------------
           0(D)       102656            999936  9.49e-05         0.0138
           1(D)        21239           1618371   0.00359          0.037
           2(D)         4008            696376    0.0433         0.0157
           3(D)          394             42796     0.276       0.000968
         --------------------------------------------------------------
         Grid Complexity: 1.24978
         Operator Complexity: 3.35769
         Total Memory Usage: 0.0674993 GB
         --------------------------------------------------------------
           iter      Mem Usage (GB)       residual           rate
         --------------------------------------------------------------
            Ini            0.880859   2.017762e+02
              0            0.880859   3.518586e+01         0.1744
              1              0.8809   8.178395e+00         0.2324
              2              0.8809   1.805407e+00         0.2208
              3              0.8809   3.697599e-01         0.2048
              4              0.8809   8.483917e-02         0.2294
              5              0.8809   1.795062e-02         0.2116
              6              0.8809   3.894573e-03         0.2170
              7              0.8809   8.352405e-04         0.2145
              8              0.8809   1.796172e-04         0.2150
              9              0.8809   3.917901e-05         0.2181
             10              0.8809   8.490560e-06         0.2167
             11              0.8809   1.810263e-06         0.2132
             12              0.8809   3.918177e-07         0.2164
             13              0.8809   8.355717e-08         0.2133
             14              0.8809   1.788109e-08         0.2140
             15              0.8809   3.816057e-09         0.2134
             16              0.8809   8.168584e-10         0.2141
             17              0.8809   1.747712e-10         0.2140
         --------------------------------------------------------------
         Total Iterations: 18
         Avg Convergence Rate: 		         0.2137
         Final Residual: 		   1.747712e-10
         Total Reduction in Residual: 	   8.661632e-13
         Maximum Memory Usage: 		          0.881 GB
         --------------------------------------------------------------
Total Time: 0.0764986
    setup: 0.0531165 s
    solve: 0.0233821 s
    solve(per iteration): 0.001299 s

|| F_h - F ||_{L^2} = 0.113514
```

Regarding the warnings: 
``CUDA Hook Library: Failed to find symbol mem_find_dreg_entries, ./amgx_ex_orig: undefined symbol: __PAMI_Invalidate_region``

Can you try running like this: ``lrun -n 1 --smpiargs="off" ./amgx_ex_orig `` ? 

Also, amgx_ex_org.cpp is the example you contributed, if there are features we would like to include into the solver class I can certainly add them in. 

Lastly, regarding running example 4p on the device. I haven't looked into running amgx with mfem on the device yet, but it looks like the following kernel has not been ported

 SparseMatrix::AddMultTranspose(line 720) 
```
     MFEM_VERIFY(Device::IsDisabled(), "transpose action on device is not "                                                                       
                  "enabled; see BuildTranspose() for details.");  
```
I'm not sure if this is not supposed to be ported? Maybe we can make this always run on the host? 

I'll work on addressing reviewer comments this week, and clean up the results as well as include expected output.

barker29 commented:
> amgx_ex_org.cpp is the example you contributed

Well, that explains why it doesn't work then...

I'll take a closer look at this - I am using the most recent pull of your branch, but I'm doing some exotic things with the configuration (xlc, essl for lapack, etc) and I'll try a more basic build and get back to you. I also missed the example runs in the PR description, so thanks for pointing those out.

artv3 commented on linalg/AmgXSolver.hpp:
```diff
@@ -0,0 +1,228 @@
+#include "sparsemat.hpp"
+#endif
+
+# define CHECK(call)                                                        \
```
removed

artv3 commented on linalg/AmgXSolver.hpp:
```diff
@@ -0,0 +1,228 @@
+
+   void Initialize_Serial(const std::string &modeStr, const std::string &cfgFile);
+
+   void SetA(const SparseMatrix &A);
```
Done!

artv3 commented on linalg/AmgXSolver.hpp:
```diff
@@ -0,0 +1,228 @@
+
+   int getNumIterations();
+
+   enum AMGX_MODE {SOLVER, PRECONDITIONER};
```
It kinda is a flag though, we can use a different convection if it is preferred.

artv3 commented on linalg/AmgXSolver.cpp:
```diff
@@ -0,0 +1,846 @@
+// \implements AmgXSolver::finalize
+void AmgXSolver::finalize()
+{
+   // skip if this instance has not been initialized
```
I added the MFEM_VERIFY, but I'm not sure if I did what you had in mind.

dylan-copeland replied:
This looks good, but wouldn't it make sense to have `if (!isInitialized || count < 1)`. In other words, shouldn't it verify that `count > 0`?

artv3 commented on linalg/AmgXSolver.cpp:
```diff
@@ -0,0 +1,846 @@
+   // Let AmgX handle returned error codes internally
+   AMGX_SAFE_CALL(AMGX_config_add_parameters(&cfg, "exception_handling=1"));
+
+   // Create an AmgX resource object, only the first instance is in charge
```
Ah, only the first instance needs to set up the library. Fixed.

tzanio commented:
This is still active, so I'm extending the review deadline to the end of this week...

artv3 commented:
@barker29 @dylan-copeland @acfisher ,  I reorganized the code a bit more and moved examples to an amgx example folder. Additionally, I included the output of the sample runs (listed on the PR description) in the amgx example folder  as a reference to the expected output.

dylan-copeland commented on examples/amgx/amgx_ex1.cpp:
```diff
@@ -0,0 +1,281 @@
+   if (!pa)
+   {
+
+      //SparseMatrix A;
```
Remove?

dylan-copeland commented on examples/amgx/amgx_ex1.cpp:
```diff
@@ -0,0 +1,281 @@
+   else // Jacobi preconditioning in partial assembly mode
+   {
+
+      //OperatorPtr A;
```
Remove?

dylan-copeland commented on linalg/amgxsolver.cpp:
```diff
@@ -0,0 +1,880 @@
+         //Bind vectors to A
+         AMGX_vector_bind(AmgXP, AmgXA);
+         AMGX_vector_bind(AmgXRHS, AmgXA);
+
```
Extra white space

dylan-copeland reviewed (approved):
Thanks @artv3 ! I still think it would be nice to have a constructor for the solver that does not require reading parameters from a file, if it is not too difficult. In any case, this PR is a very useful contribution.

barker29 commented on examples/amgx/makefile:
```diff
@@ -0,0 +1,72 @@
+-include $(CONFIG_MK)
+
+SEQ_EXAMPLES = amgx_ex1
+PAR_EXAMPLES = amgx_ex1p amgx_ex4p
```
`amgx_ex4p.cpp` appears to have disappeared from the repo?

artv3 replied:
I ended up removing it since some of the mfem routines had blocks  when CUDA was used.

barker29 commented on linalg/amgxsolver.hpp:
```diff
@@ -0,0 +1,220 @@
+   AmgXSolver() = default;
+
+   /* Constructor for serial builds - supported without Hypre and MPI */
+   AmgXSolver(const std::string &cfgFile);
```
As other reviewers have suggested, I think an interface that does not require a configuration file would be useful. We can follow roughly the thinking behind `HypreBoomerAMG`, that is, unless you specify otherwise it configures as a preconditioner that does one V-cycle with some standard options.

barker29 commented on linalg/amgxsolver.hpp:
```diff
@@ -0,0 +1,220 @@
+   // Indicate if this instance has been initialized.
+   bool                    isInitialized = false;
+
+#ifdef MFEM_USE_MPI
```
The amount of effort that went into making this work in the multi-GPU setting is impressive!

barker29 commented on linalg/amgxsolver.cpp:
```diff
@@ -0,0 +1,880 @@
+                                     const Array<int64_t> &loc_J,
+                                     const bool update_mat)
+{
+   // The following arrays hold the
```
Incomplete comment?

barker29 commented on linalg/amgxsolver.cpp:
```diff
@@ -0,0 +1,880 @@
+      AMGX_solver_get_status(solver, &status);
+      if (status != AMGX_SOLVE_SUCCESS && amgxMode == SOLVER)
+      {
+         printf("Amgx failed to solve system, error code %d. \n", status);
```
Something like `mfem_error` or `MFEM_VERIFY` probably makes more sense here than a printf.

barker29 commented on linalg/amgxsolver.cpp:
```diff
@@ -0,0 +1,880 @@
+      AMGX_solver_get_status(solver, &status);
+      if (status != AMGX_SOLVE_SUCCESS && amgxMode == SOLVER)
+      {
+         printf("Amgx failed to solve system, error code %d. \n", status);
```
Ditto.

barker29 commented on setupAMGXMFEM.sh:
```diff
@@ -0,0 +1,18 @@
+#cd hypre/src/ && mkdir build && cd build &&
+#cmake -DCMAKE_INSTALL_PREFIX=$(pwd)/../../ ../ && make -j && make install &&
+
+cd ../ &&
```
This file is pretty helpful but probably doesn't belong in the repo. Maybe a very brief instruction on installing AMGx could be put in the changelog or readme?

artv3 replied:
deal!

tzanio replied:
This should probably go in `INSTALL`

barker29 commented on linalg/amgxsolver.cpp:
```diff
@@ -0,0 +1,880 @@
+   Array<int64_t> loc_J((int)A_csr->num_nonzeros);
+   for (int i=0; i<A_csr->num_nonzeros; ++i)
+   {
+      loc_J[i] = A_csr->big_j[i];
```
I had to replace `big_j` with `j` to get this to compile. Does this depend on the hypre version you are using?

artv3 replied:
I've been grabbing the version from master: `` https://github.com/hypre-space/hypre.git``. If it doesn't work with older versions I can make a note. But the column index is expected to be of int64_t type.

barker29 replied:
`big_j` was introduced between hypre 2.15.1 and 2.16 (https://github.com/hypre-space/hypre/blob/fe8a0e0e0b1c54299f804726990d2942ae505e34/src/seq_mv/csr_matrix.h#L31). We can probably require the newer hypre for AmgX but it should be noted somewhere.

Unrelatedly, does it make sense to use `HYPRE_BigInt` instead of `int64_t`?

artv3 replied:
I think ``HYPRE_BigInt`` has to be enabled to be int64 right? If it always is, then yes we can do that.

barker29 replied:
We should either do some hypre version include guard here, or document carefully that our AMGx support also requires hypre >= 2.16.

barker29 reviewed (changes requested):
I am very excited about this capability and am eager to use it.

That said, the PR in its current form needs a little bit of polish before it can be merged into MFEM. I'd like to see a cleaner and easier-to-use interface to `AmgXSolver` and also an example for non-experts, where they can just use AmgX on device by changing a couple lines of code.

dylan-copeland commented on examples/amgx/amgx_ex1.cpp:
```diff
@@ -0,0 +1,281 @@
+//                                MFEM Example 1
```
I am also in favor of a simple option in ex1p, as I have noted in comments before. With support now from more than 1 reviewer, it seems this should be done.

artv3 replied:
Deal!

artv3 replied:
@dylan-copeland @barker29 , given that AMGX supports many solvers/preconditioners. Is there a preference for default solver, preconditioner?  I was thinking for default solver using the amg_pcg.json configuration.

barker29 replied:
I think the default should be close to what `HypreBoomerAMG` does, that is, a single V-cycle acting as a preconditioner. I may be wrong but I think most MFEM users will view AmgX as a preconditioner and wrap it with MFEM solvers?

tzanio replied:
Agreed

artv3 commented on examples/amgx/SampleOutput.md:
```diff
@@ -0,0 +1,551 @@
+
+Sample 4
+
+lrun -n 4 ./amgx_ex1p --amgx-file amg_pcg.json --amgx-mpi-gpu-exclusive
```
Hmh.... This is strange, I just built and ran this example...

barker29 replied:
Never mind, this was my mistake in configuring amgx.

artv3 commented on examples/amgx/amgx_ex1p.cpp:
```diff
@@ -0,0 +1,318 @@
+
+      amgx.SetOperator(*A.As<HypreParMatrix>());
+
+      amgx.Mult(B, X);
```
Depends on how the amgx is configured, with the sample run it will completely solve.

barker29 commented on examples/amgx/SampleOutput.md:
```diff
@@ -0,0 +1,551 @@
+Converting config string to current config version
+Parsing configuration string: exception_handling=1 ; 
+Using Normal MPI (Hostbuffer) communicator...
+AMG Grid:
```
This is probably more related to AMGx than your PR, but is there a way to not duplicate this output on every processor? It's especially annoying in the iteration monitor output below.

artv3 replied:
I'll look around and see if we can do anything about that.

artv3 commented:
Thanks for the feedback guys, I didn't get to them during the week. But I'll try to get the PR in shape by Monday.

tzanio commented:
Adding a bit more time for review as there are still changes...

artv3 commented:
@barker29 @dylan-copeland  I added a default configuration for using AmgX as a preconditioner or solver and added example usage under example 1/1p. I still have to do a proof read pass and do minor clean up. But I was wondering if examples 1/1p illustrate the usage we are looking for? Using the amgx option is activated using the -amgx flag as illustrated below. The amgx folder contains additional examples for other features. 

The output one could expect from ex1/1p using AmgX is illustrated below. 
```
lrun -n 1 --smpiargs="off" ./ex1 -amgx

Compiled with CUDA Runtime 10.1, using CUDA driver 10.2
   Iteration :   0  (B r, r) = 0.685373
   Iteration :   1  (B r, r) = 9.47709e-05
   Iteration :   2  (B r, r) = 1.15799e-07
   Iteration :   3  (B r, r) = 2.31355e-10
   Iteration :   4  (B r, r) = 4.80418e-13
Average reduction factor = 0.030249
```

```
lrun -n 2 ./ex1p -amgx 

Options used:
   --mesh ../data/star.mesh
   --order 1
   --no-static-condensation
   --no-partial-assembly
   --amgx-precon
   --device cpu
   --visualization
Device configuration: cpu
Memory configuration: host-std
Number of finite element unknowns: 82561
AMGX version 2.1.0.131-opensource
Built on Oct  2 2020, 14:50:59
Compiled with CUDA Runtime 10.1, using CUDA driver 10.2
Cannot read file as JSON object, trying as AMGX config
Converting config string to current config version
Parsing configuration string: exception_handling=1 ; 
Using Normal MPI (Hostbuffer) communicator...
   Iteration :   0  (B r, r) = 0.387334
   Iteration :   1  (B r, r) = 0.0730444
   Iteration :   2  (B r, r) = 0.000394756
   Iteration :   3  (B r, r) = 1.05156e-05
   Iteration :   4  (B r, r) = 5.62496e-08
   Iteration :   5  (B r, r) = 2.52817e-10
   Iteration :   6  (B r, r) = 8.3872e-12
   Iteration :   7  (B r, r) = 2.19847e-13
   Iteration :   8  (B r, r) = 8.09957e-16
   Iteration :   9  (B r, r) = 6.98018e-18
   Iteration :  10  (B r, r) = 1.99383e-20
   Iteration :  11  (B r, r) = 3.32709e-23
   Iteration :  12  (B r, r) = 6.47468e-26
Average reduction factor = 0.0928176
```

artv3 commented on examples/amgx/SampleOutput.md:
```diff
@@ -0,0 +1,551 @@
+Sample 1
+
```
TODO: Remove...

barker29 commented on examples/ex1.cpp:
```diff
@@ -191,7 +194,31 @@ int main(int argc, char *argv[])
+   {
+#if defined(MFEM_USE_AMGX)
+      bool amgx_verbose = false;
+      AmgXSolver amgx(AmgXSolver::PRECONDITIONER, amgx_verbose);
```
This is exactly the kind of interface I was hoping for, very nice.

barker29 commented on examples/ex1p.cpp:
```diff
@@ -84,6 +85,8 @@ int main(int argc, char *argv[])
    args.AddOption(&pa, "-pa", "--partial-assembly", "-no-pa",
                   "--no-partial-assembly", "Enable Partial Assembly.");
+   args.AddOption(&amgx, "-amgx", "--amgx-precon", "-no-amgx",
+                  "--no-amgx-precon", "Use AmgX.");
```
A slightly longer help line would be nice, maybe "Use AmgX V-cycle as preconditioner for CG."

barker29 commented on linalg/amgxsolver.hpp:
```diff
@@ -0,0 +1,227 @@
+
+   AmgXSolver() = default;
+
+   /* Constructor for serial builds - supported without Hypre and MPI */
```
These comment lines should be doxygen-ized.

barker29 commented on linalg/amgxsolver.hpp:
```diff
@@ -0,0 +1,227 @@
+   AmgXSolver(const MPI_Comm &comm, const int nDevs,
+              const AMGX_MODE amgx_Mode_, const bool verbose);
+
+   void Initialize_ExclusiveGPU(const MPI_Comm &comm);
```
For the next few methods, I suggest adding (doxygen) comment lines to document the purpose of these methods and also changing the style to be more MFEM-y, eg `InitializeExclusiveGPU`.

barker29 commented on linalg/amgxsolver.hpp:
```diff
@@ -0,0 +1,227 @@
+
+public:
+
+   enum AMGX_MODE {SOLVER, PRECONDITIONER};
```
Here (or maybe before the `DefaultParameters` method you should document what these defaults are, at least the basic pionts (ie, which smoother?)

barker29 commented on linalg/amgxsolver.hpp:
```diff
@@ -0,0 +1,227 @@
+namespace mfem
+{
+
+class AmgXSolver : public Solver
```
This class deserves a pretty detailed comment block describing its purpose and use. Maybe even also a link to the AmgX github page and a very brief outline of how to configure MFEM with AmgX?

barker29 commented on linalg/amgxsolver.cpp:
```diff
@@ -0,0 +1,1026 @@
+      AMGX_solver_setup(solver, AmgXA);
+      AMGX_vector_bind(AmgXP, AmgXA);
+      AMGX_vector_bind(AmgXRHS, AmgXA);
+
```
As a very minor point, this file has a lot of these blank lines that are not typical of MFEM style.

barker29 reviewed (approved):
I'm happy with this PR and think it will be very useful. All my comments are minor or superficial.

tzanio commented:
@v-dobrev, do you want to review this?

v-dobrev commented:
> @v-dobrev, do you want to review this?

I think we have enough approvals. However, I noticed that not all reviewer comments have been addressed, e.g. https://github.com/mfem/mfem/pull/1386#discussion_r496233978.

artv3 commented:
Travis is failing because of my documentation style, is there a way to test locally if I am documenting with the right format?

tzanio commented:
You can likely just do `make` in the `doc/` directory. If you don't have Doxygen, you will have to install it before that.

That being said, in this case the error seems clear:
```
/home/travis/build/mfem/mfem/linalg/amgxsolver.hpp:171: warning: argument 'AMGX_MODE' of command @param is not found in the argument list of mfem::AmgXSolver::DefaultParameters(const AMGX_MODE amgxMode_, const bool verbose)
/home/travis/build/mfem/mfem/linalg/amgxsolver.hpp:185: warning: The following parameters of mfem::AmgXSolver::DefaultParameters(const AMGX_MODE amgxMode_, const bool verbose) are not documented:
  parameter 'amgxMode_'
```

artv3 commented:
I believe I have addressed all the comments, one thing I haven't been able to figure out is how to build with CMAKE....

tzanio commented:
Merged in `next` for testing...

v-dobrev commented:
Re-merged in `next`.

tzanio commented:
Re-merged in `next` for testing ...

dylan-copeland commented:
I have not done or seen timing comparisons for AMGX and hypre until now, but I just ran a test for `ex1p -m ../data/inline-hex.mesh` with and without `-amgx`, with 4 MPI tasks. Hypre took 0.53 seconds with 24 iterations; AMGX took 0.079 seconds with 10 iterations.

barker29 commented:
@dylan-copeland are they both using GPU?

v-dobrev commented on examples/ex1.cpp:
```diff
@@ -82,6 +87,8 @@ int main(int argc, char *argv[])
                   "--no-static-condensation", "Enable static condensation.");
    args.AddOption(&pa, "-pa", "--partial-assembly", "-no-pa",
                   "--no-partial-assembly", "Enable Partial Assembly.");
+   args.AddOption(&amgx, "-amgx", "--amgx-precon", "-no-amgx",
+                  "--no-amgx-precon", "Use AmgX V-cycle as preconditioner for CG.");
```
I'm not sure if this was discussed before: since we have a separate set of `ex1`/`ex1p` in the `amgx` sub-directory, isn't it better to keep the regular `ex1`/`ex1p` simpler and remove the AmgX options from them?

tzanio commented:
Re-merged in `next` for testing ...

v-dobrev commented on linalg/amgxsolver.cpp:
```diff
@@ -269,6 +269,12 @@ void AmgXSolver::InitAmgX()
+                        [](const char *msg, int length)->void
+      {
+         int irank; MPI_Comm_rank(MPI_COMM_WORLD, &irank);
+         if (irank == 0) { printf("%s",msg);} }));
```
Use `mfem::out` for output instead of `std::cout`/`printf`.

tzanio commented:
Re-merged in `next` for testing ...

artv3 commented:
There was a hypre error with Travis with the current configuration:

Mac: Parallel + Debug
AMD64
Compiler: clang Xcode: xcode11.2 C++ 

```
linalg/cpardiso.cpp -o linalg/cpardiso.o

In file included from linalg/cpardiso.cpp:2:

In file included from linalg/hypre.hpp:25:

In file included from ./../hypre/src/hypre/include/seq_mv.h:13:

./../hypre/src/hypre/include/HYPRE_seq_mv.h:17:10: fatal error: 'HYPRE_utilities.h' file not found

#include "HYPRE_utilities.h"
```

artv3 commented:
> There was a hypre error with Travis with the configuration:
> 
> Mac: Parallel + Debug
> AMD64
> Compiler: clang Xcode: xcode11.2 C++
> 
> ```
> linalg/cpardiso.cpp -o linalg/cpardiso.o
> 
> In file included from linalg/cpardiso.cpp:2:
> 
> In file included from linalg/hypre.hpp:25:
> 
> In file included from ./../hypre/src/hypre/include/seq_mv.h:13:
> 
> ./../hypre/src/hypre/include/HYPRE_seq_mv.h:17:10: fatal error: 'HYPRE_utilities.h' file not found
> 
> #include "HYPRE_utilities.h"
> ```

Fixed after clearing cache and restarting build, thanks @v-dobrev!

tzanio commented:
Re-merged in `next` for testing ...

tzanio reviewed (approved):
This has to be one of my favorite PRs in a long time... Thank you so much for working on this @artv3! 🚀 

@dylan-copeland, @acfisher, @v-dobrev, @barker29 and @mattmartineau -- thank you for your help and reviews!

artv3 commented:
> I have not done or seen timing comparisons for AMGX and hypre until now, but I just ran a test for `ex1p -m ../data/inline-hex.mesh -d cuda` with and without `-amgx`, with 4 MPI tasks on 4 GPUs. Hypre took 0.48 seconds with 24 iterations; AMGX took 0.049 seconds with 10 iterations. This is using hypre 2.20.0 built with cuda and UVM.

It would be good to compare what the configuration for each of the solvers. If we get them the same the number of iterations should be closer.

artv3 commented:
@tzanio @dylan-copeland , currently with the AmgX defaults, using AmgX as a preconditioner will invoke two iterations of a v-cycle with a Jacobi smoother. I think Hypre may only do one. In general by doing more v-cycles on the GPU I found better performance in both run-time and number of iterations. Should we keep two as the default? or change to one?

tzanio commented:
Yes Hypre does one v-cycle. On the CPU that typically is the best choice for overall time to solution.

I think it is OK to leave it at 2 (or more) on the GPU, but it will be good to have some benchmarks to see that it is better, and we should probably also document it in the code.
